openapi.yaml

```diff
@@ -0,0 +1,2411 @@
+
+      responses:
+        '200':
+          description: Successful response.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/XCom'
+        '400':
+          $ref: '#/components/responses/BadRequest'
+        '401':
+          $ref: '#/components/responses/Unauthenticated'
+        '403':
+          $ref: '#/components/responses/PermissionDenied'
+        '404':
+          $ref: '#/components/responses/NotFound'
+
+    delete:
+      summary: Delete an XCom entry
+      operationId: deleteXComValue
+      tags: [XCom]
+      responses:
+        '204':
+          description: No content.
+        '400':
+          $ref: '#/components/responses/BadRequest'
+        '401':
+          $ref: '#/components/responses/Unauthenticated'
+        '403':
+          $ref: '#/components/responses/PermissionDenied'
+
+  # Non-database resources
+  /dags/{dag_id}/dagRuns/{execution_date}/taskInstances/{task_id}/links:
+    parameters:
+      - $ref: '#/components/parameters/DAGID'
+      - $ref: '#/components/parameters/ExecutionDate'
+      - $ref: '#/components/parameters/TaskID'
+
+    get:
+      summary: Get extra links for task instance
+      operationId: getExtraLinks
+      tags: [TaskInstance]
+      responses:
+        '200':
+          description: Successful response.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ExtraLinkCollection'
+        '401':
+          $ref: '#/components/responses/Unauthenticated'
+        '403':
+          $ref: '#/components/responses/PermissionDenied'
+        '404':
+          $ref: '#/components/responses/NotFound'
+
+  /dags/{dag_id}/dagRuns/{execution_date}/taskInstances/{task_id}/logs/{task_try_number}:
+    parameters:
+      - $ref: '#/components/parameters/DAGID'
+      - $ref: '#/components/parameters/ExecutionDate'
+      - $ref: '#/components/parameters/TaskID'
+      - $ref: '#/components/parameters/TaskTryNumber'
+      - $ref: '#/components/parameters/NoChunking'
+      - $ref: '#/components/parameters/ContinuationToken'
+
+    get:
+      summary: Get logs for specific task instance
+      operationId: getLogs
+      tags: [TaskInstance]
+      responses:
+        '200':
+          description: Content of logs.
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  continuation_token:
+                    type: string
+                  content:
+                    type: string
+            text/plain:
+              schema:
+                type: string
+        '400':
+          $ref: '#/components/responses/BadRequest'
+        '401':
+          $ref: '#/components/responses/Unauthenticated'
+        '403':
+          $ref: '#/components/responses/PermissionDenied'
+        '404':
+          $ref: '#/components/responses/NotFound'
+
+  /dags/{dag_id}/structure:
```

Comment:
IIUC, this endpoint is separate from `/dags/{dag_id}` because of different sources of data. `/dags/{dag_id}` returns data present in the database while `/structure` returns data retrieved from the DAG python code.

@kaxil
Is my thinking correct that we could have only one endpoint if DAG serialization was always turned on? If so, what is the plan for DAG serialization in Airflow 2.0? Is it going to be configurable or is it going to be always on?

Reply:
In Airflow 2.0, we want to introduce DAG versioning, so connecting these two endpoints will be rather difficult. In the future, we will be adding new endpoints to provide access to historical versions of DAG structures. `/dags/{dag_id}/` and `/dags/` will provide general metadata e.g. is_paused. 

Is there a reason why merging these two endpoints is important to you?

Reply:
I was thinking about making the API simpler by having one fewer endpoint.

I notice that there are some fields that appear in both `DAG` and `DagStructure` (`schedule_interval`, `is_subdag`, `fileloc`). My guess is that the reason is that these fields have been "serialized" from Python to the database and the other fields were not. Maybe an alternative to merging the 2 endpoints would be to make them disjoint and remove the duplicated fields from `DAG`?

Maybe this is a good moment to start discussing how the DAG versioning API will look like?

Reply:
We need these fields to effectively build a user interface.
![https://airflow.readthedocs.io/en/latest/_images/dags.png](https://airflow.readthedocs.io/en/latest/_images/dags.png)
`schedule_interval` is used by home page.
`is_subdag` is used by filter. Home page doesn't display subDAGs..
`fileloc` is important because it is used as an identifier at another endpoint. - `/dagSources/{file_token}:`. This is not yet clearly visible in this specification, because HATEOS will be implemented after accepting the initial specification, but HATEOS will use this parameter.

From what I know, there has already been one meeting on this issue and one person from your team was present. For now, the community is focusing on improvements to DAG serialization and in the next step using DAG serialization in a scheduler to provide [High Avaliblity](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=103092651) . I think that DAG versioning is suspended because we want to divide it into two phases. DAG versioning for Web UI, and then DAG versioning for the scheduler. However, Web UI is in bad condition, and the way to improve its condition is to use the API together with [React](https://lists.apache.org/thread.html/653052f7e5f72af4d22f155ddaf48b855d496b9070488bc2b668810d%40%3Cdev.airflow.apache.org%3E) to build UI.  Once we have the API and improvements for Web UI, we can plan work on DAG versioning
Summarizing. 
REST API => React UI => DAG Versioning
DAG Serialization for Web UI => DAG Serialization for scheduler/HA => DAG Versioning